spring/spring6/spring6-common/src/main/java/org/infinispan/spring/common/provider/SpringCache.java

```diff
@@ -259,4 +267,49 @@ static RuntimeException throwValueRetrievalException(Object key, Callable<?> loa
    public long getWriteTimeout() {
       return writeTimeout;
    }
+
+   @Override
+   public CompletableFuture<?> retrieve(Object key) {
+      return encodedToValueWrapper(nativeCache.getAsync(key));
+   }
+
+   @Override
+   public <T> CompletableFuture<T> retrieve(Object key, Supplier<CompletableFuture<T>> valueLoaderAsync) {
+      CompletionStage<T> completionStage = CompletionStages.handleAndCompose(nativeCache.getAsync(key), (v1, ex1) -> {
+         if (ex1 != null) {
+            return CompletableFuture.failedFuture(ex1);
+         }
+
+         if (v1 != null) {
+            return CompletableFuture.completedFuture(decodeNull(v1));
+         }
+
+         CompletableFuture<T> computedValue = computationResults.putIfAbsent(key, new CompletableFuture<>());
+         if (computedValue != null) {
+            return computedValue;
+         }
+         return CompletionStages.handleAndCompose(valueLoaderAsync.get(), (newValue, ex3) -> {
+                     CompletableFuture<T> result = computationResults.get(key);
+                     if (ex3 != null) {
+                        result.completeExceptionally(ex3);
+                        computationResults.remove(key);
+                     } else {
+                        CompletionStages.handleAndCompose(
+                              nativeCache.putIfAbsentAsync(key, encodeNull(newValue)), (existing, ex4) -> {
```

Comment:
Just nitpick, we can change this to a `nativeCache.putIfAbsentAsync(key, encodeNull(newValue)).whenComplete((existing, ex4) -> ...)`, without needing the `CompletionStages.handleAndCompose`

Reply:
Crazy.. I had this in my review but I seem to have lost it. Yeah, if you aren't using the return value of handleAndCompose, then it isn't needed.